03-task-manager/starter/QuizAnswers2.txt

```diff
@@ -0,0 +1,20 @@
+    In this lesson, you created a middleware function called asyncWrapper. Why?
+    
+    We want async functions because they are nonblocking. 
+    However, adding a try catch block to each function is repetitive and long.
+    In order to prevent this repetitive code we wrap the asynchronous try/catch
+    into a middleware function that will perform the try catch and then return the value.
+
+    Suppose that you want to make sure that both a status code and and error message are send back 
+    to the user when they request the URL for a task that does not exist. 
+    Assume that you’ve created a CustomAPIError class and an error handler that references that class. 
+    Complete the code:
+
+const getTask = asyncWrapper(async (req, res, next) => {
+  const { id: taskID } = req.params;
+  const task = await Task.findOne({ _id: taskID });
+  if (!task) {
+    return next(CustomAPIError(`No task with id : ${taskID}`, 404))
```

Comment:
Almost! Because `CustomAPIError` is a class, you need to instantiate it with `new`.

Reply:
Okay, so just to make sure I understand what I am doing... I need to do new because if there's no matching task the error is an instance of the CustomAPIError class. A lot of the examples I saw when I was looking up errors in node/express involved using throw which I wouldn't want to do because that stops anymore functions from running if there's an error and just goes to the catch statement if it exists, right? I didn't see many examples with next. Can I just add the word new in front of CustomAPIError like this?
```
const getTask = asyncWrapper(async (req, res, next) => {
  const { id: taskID } = req.params;
  const task = await Task.findOne({ _id: taskID });
  if (!task) {
    return next(new CustomAPIError(`No task with id : ${taskID}`, 404))
  }
  res.status(200).json({ task });
});
```

or I have to instantiate the class by adding a function like this?
 ```
const newCustomError = (message, status) => { return new CustomAPIError(message, status)}

const getTask = asyncWrapper(async (req, res, next) => {
  const { id: taskID } = req.params;
  const task = await Task.findOne({ _id: taskID });
  if (!task) {
    return next(newCustomAPIError(`No task with id : ${taskID}`, 404))
  }
  res.status(200).json({ task });
});
```
or I am totally off?

Reply:
You're heading in the right direction!

1. Whenever you have a `class` in JavaScript, you have to _instantiate_ it (i.e. a class is a type of object and you want to create on) by using the *new* keyword => `new CustomAPIError()`. [The MDN docs are pretty good on this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes). This is what is happening in the `errors/custom-error.js` file [in the `createCustomError` function on line 9](https://github.com/JenMcD-star/node-express-course/blob/6aa34bd4fa318184289e686e0ba13fbeb44a713f/03-task-manager/starter/errors/custom-error.js#L9).

2. In Express v4, you are right that throwing inside of an async function will exit the request handling and is bad. This is why we created the [wrapper method](https://github.com/JenMcD-star/node-express-course/blob/6aa34bd4fa318184289e686e0ba13fbeb44a713f/03-task-manager/starter/middleware/async.js) that allows you to throw in the main logic and catches those errors calling `next` with them. [There are some examples in the Express error-handling guide](https://expressjs.com/en/guide/error-handling.html).